dockstore-webservice/src/main/java/io/dockstore/webservice/helpers/QuayImageRegistry.java

```diff
@@ -157,83 +176,92 @@ public List<Tag> getTags(Tool tool) {
+    public Set<Image> handleMultiArchQuayTags(String repository, QuayTag quayTag, List<QuayTag> cleanedQuayTagList, LanguageHandlerInterface.DockerSpecifier specifier) throws ApiException {
         QuayRepoManifest quayRepoManifest;
         DockerManifestList manifestList;
-        List<Image> images = new ArrayList<>();
-        quayRepoManifest = manifestApi.getRepoManifest(quayTag.getManifestDigest(), tool.getNamespace() + '/' + tool.getName());
+        Set<Image> images = new HashSet<>();
+        quayRepoManifest = manifestApi.getRepoManifest(quayTag.getManifestDigest(), repository);
         try {
-            manifestList = g.fromJson(quayRepoManifest.getManifestData(), DockerManifestList.class);
+            manifestList = GSON.fromJson(quayRepoManifest.getManifestData(), DockerManifestList.class);
         } catch (JsonSyntaxException ex) {
-            LOG.info("Unexpected response from Quay while retrieving repo manifest for Quay Tag " + quayTag.getName() + " for tool " + tool.getToolPath(), ex);
+            LOG.info("Unexpected response from Quay while retrieving repo manifest for Quay Tag " + quayTag.getName() + " for repository " + repository, ex);
             return images;
         }
+
         try {
+            // Create an image for each architecture manifest
             Arrays.stream(manifestList.getManifests()).forEach(manifest -> {
-                Optional<QuayTag> matchingTag = cleanedQuayTagList.stream().filter(tag -> (tag.getManifestDigest().equals(manifest.getDigest()))).findFirst();
-                if (matchingTag.isPresent()) {
-                    final String manifestDigest = manifest.getDigest();
-                    final String imageID = matchingTag.get().getImageId();
-                    List<Checksum> checksums = new ArrayList<>();
-                    String[] splitChecksum = manifestDigest.split(":");
-                    checksums.add(new Checksum(splitChecksum[0], splitChecksum[1]));
-                    Image image = new Image(checksums, tool.getNamespace() + '/' + tool.getName(), quayTag.getName(), imageID, Registry.QUAY_IO, quayTag.getSize(), quayTag.getLastModified());
-                    image.setArchitecture(manifest.getPlatform().getArchitecture());
-                    image.setOs(manifest.getPlatform().getOs());
-                    images.add(image);
-                    cleanedQuayTagList.remove(matchingTag.get());
-                } else {
-                    LOG.info("Unable to find matching Quay tag image information using digest");
-                }
+                final String manifestDigest = manifest.getDigest();
+                final String imageID = quayTag.getImageId();
+                List<Checksum> checksums = new ArrayList<>();
+                String[] splitChecksum = manifestDigest.split(":");
+                checksums.add(new Checksum(splitChecksum[0], splitChecksum[1]));
```

Comment:
To be super safe, we could first check that `manifestDigest` has two parts here.  Could even make a method like `createChecksumFromManifest` that encapsulates that logic and logs/throws if it's passed something weird, and use here and at line 260.  Yes, all this checking is a bit paranoid, but would help us to diagnose if Quay changes their manifest format or there's a bug...  Or maybe the manifest has already been checked elsewhere?

Reply:
Hmm, it hasn't been checked, but this [manifest format](https://docs.docker.com/registry/spec/manifest-v2-2/#manifest-list) is established by Docker, not Quay, if that's any better. Should I still do the checking?